### DISCUSSION TITLE
MailCow as oAuth Provider

### DISCUSSION THREAD

**Discussion by kilian099** [nextauthjs/next-auth#9206] [Help]
Hi All, I wanted to express my interest in learning about NextAuth (I am a bit newbie on this kind of backends with JSON. I've got as email provider it's called mailcow and they have their own API connection for oAuth Authorization to logon into the server. However, I am working through in a custom app where I would like to implement the logins through that server and create profiles in this app with the Email provider details. Could you please can tell me how I should start to request callbacks and also what provide should I import for the login page? Obviously, I cannot see MailCow as a provider but I guess I can use different provider for built this custom login. Is that correct?

**Comment by rallisf1** [comment]
a little late to the party but got it working after an hour of fiddling

```js
        {
            id: 'mailcow',
            name: 'Mailcow Authentication',
            type: 'oauth',
            token: 'https://your.mailserver.domain/oauth/token',
            authorization: { url: 'https://your.mailserver.domain/oauth/authorize', params: { scope: 'profile' } },
            checks: ["state"],
            style: {
              logo: 'https://avatars.githubusercontent.com/u/23747925?s=48&v=4',
            },
            userinfo: 'https://your.mailserver.domain/oauth/profile',
            clientId: MAILCOW_CLIENT_ID,
            clientSecret: MAILCOW_CLIENT_SECRET,
            async profile(profile){
              return {
                  name: profile?.full_name,
                  email: profile?.email,
              }
            }
        }
```

I might create a PR for it later.
